client/client.go

```diff
@@ -109,71 +103,86 @@ func (c *LogClient) ListByIndex(ctx context.Context, start, count int64) ([]*tri
+			if root.TreeSize >= waitForTreeSize {
+				return root, nil
+			}
+		case codes.Unavailable: // Retry.
+		default:
+			return nil, err
 		}
-		if err := ctx.Err(); err != nil {
-			return status.Errorf(codes.DeadlineExceeded,
-				"%v. TreeSize: %v, want > %v. Tried %v times.",
-				err, c.root.TreeSize, startTreeSize, i+1)
+
+		select {
+		case <-ctx.Done():
+			return nil, status.Errorf(codes.DeadlineExceeded,
+				"%v. TreeSize: %v, want >= %v. Tried %v times: %v",
+				err, c.root.TreeSize, waitForTreeSize, i+1, ctx.Err())
+		case <-time.After(b.Duration()):
 		}
-		time.Sleep(b.Duration())
 	}
 }
 
-// UpdateRoot retrieves the current SignedLogRoot.
-// Verifies the signature, and the consistency proof if this is not the first root this client has seen.
-func (c *LogClient) UpdateRoot(ctx context.Context) error {
+// GetLatestRoot fetches and verifies the latest root against a trusted root, seen in the past.
+// Pass nil for trusted if this is the first time querying this log.
+func (c *LogClient) GetLatestRoot(ctx context.Context, trusted *trillian.SignedLogRoot) (*trillian.SignedLogRoot, error) {
 	resp, err := c.client.GetLatestSignedLogRoot(ctx,
 		&trillian.GetLatestSignedLogRootRequest{
 			LogId: c.LogID,
 		})
 	if err != nil {
-		return err
+		return nil, err
 	}
-	if c.root.TreeSize > 0 &&
-		resp.SignedLogRoot.TreeSize == c.root.TreeSize &&
-		bytes.Equal(resp.SignedLogRoot.RootHash, c.root.RootHash) {
+	if trusted.TreeSize > 0 &&
+		resp.SignedLogRoot.TreeSize == trusted.TreeSize &&
+		bytes.Equal(resp.SignedLogRoot.RootHash, trusted.RootHash) {
 		// Tree has not been updated.
-		return nil
+		return resp.SignedLogRoot, nil
 	}
 	// Fetch a consistency proof if this isn't the first root we've seen.
 	var consistency *trillian.GetConsistencyProofResponse
-	if c.root.TreeSize > 0 {
+	if trusted.TreeSize > 0 {
 		// Get consistency proof.
 		consistency, err = c.client.GetConsistencyProof(ctx,
 			&trillian.GetConsistencyProofRequest{
 				LogId:          c.LogID,
-				FirstTreeSize:  c.root.TreeSize,
+				FirstTreeSize:  trusted.TreeSize,
 				SecondTreeSize: resp.SignedLogRoot.TreeSize,
 			})
 		if err != nil {
-			return err
+			return nil, err
 		}
 	}
 	// Verify root update if the tree / the latest signed log root isn't empty.
 	if resp.GetSignedLogRoot().GetTreeSize() > 0 {
-		if err := c.logVerifier.VerifyRoot(&c.root, resp.GetSignedLogRoot(),
+		if err := c.logVerifier.VerifyRoot(trusted, resp.GetSignedLogRoot(),
 			consistency.GetProof().GetHashes()); err != nil {
-			return err
+			return nil, err
 		}
-		c.root = *resp.SignedLogRoot
 	}
-	return nil
+	return resp.SignedLogRoot, nil
+}
+
+// UpdateRoot retrieves the current SignedLogRoot, verifying it against roots this client has
+// seen in the past, and updating the currently trusted root if the new root verifies.
+func (c *LogClient) UpdateRoot(ctx context.Context) (*trillian.SignedLogRoot, error) {
+	newTrusted, err := c.GetLatestRoot(ctx, &c.root)
```

Comment:
Thinking of clock-skew, caching, etc.; `newTrusted` could be older/smaller than the currently trusted root?

Reply:
Hmm. Good thinking. Yeah, especially if we happen to talk to a different cluster.  I'll add a check